design/plugin-backup-and-restore-progress-design.md

```diff
@@ -0,0 +1,238 @@
+	// The server's time is used for CompletionTimestamps
+	// +optional
+	// +nullable
+	CompletionTimestamp *metav1.Time `json:"completionTimestamp,omitempty"`
+
+	// Progress holds the total number of bytes of the snapshot and the current
+	// number of restored bytes. This can be used to display progress information
+	// about the restore operation.
+	// +optional
+	Progress VolumeOperationProgress `json:"progress,omitempty"`
+}
+
+type VolumePluginRestore struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// +optional
+	Spec VolumePluginRestoreSpec `json:"spec,omitempty"`
+
+	// +optional
+	Status VolumePluginRestoreStatus `json:"status,omitempty"`
+}
+```
+
+For every restore operation of volume, Volume snapshotter creates VolumePluginRestore CR in Velero namespace.
+Plugin keep updating the progress of operation along with other details like Volume name, Backup Name, SnapshotID etc as mentioned in the CR.
+
+Snapshot ID that need to be restored will help to identify related VolumePluginBackup CR.
+
+In order to identify the CR created for the particular restore of a volume, volume snapshotters adds following labels to CR:
+`velero.io/backup-name` with value as Backup Name, and,
+`velero.io/snapshot-id` with value as snapshot id that need to be restored
+
+### API Upgrade
+When CRs gets upgraded, velero can support older API versions also (till they get deprecated) to identify the CRs that need to be persisted to backup location.
+However, it can provide preference over latest supported API.
+
+If new fields are added without changing API version, it won't cause any problem as these resources are intended to provide information, and, there is no reconciliation on these resources.
+
+##Limitations:
+
+Non K8s native plugins will not be able to implement this as they can not create the CRs.
+
+## Open Questions
+
+## Alternatives Considered
+
+### Add another method to VolumeSnapshotter interface 
+Above proposed approach have limitation that plugin need to be K8s native in order to create, update CRs.
+Instead, a new method for 'Progress' will be added to interface. Velero server regularly polls this 'Progress' method and updates VolumePluginBackup CR on behalf of plugin.
+
+But, this involves good amount of changes and needs a way for backward compability.
+
+As volume plugins are mostly K8s native, its fine to go ahead with current limiation.
+
+### Update Backup CR
+Instead of creating new CRs, plugins can directly update the status of Backup CR. But, this deviates from current approach of having seperate CRs like PodVolumeBackup/PodVolumeRestore to know operations progress.
+
+### Restricting on name rather than using labels
+Instead of using labels to identify the CR related to particular backup on a volume, restrictions can be placed on the name of VolumePluginBackup CR to be same as the value returned from CreateSnapshot.
```

Comment:
Note that values in labels should be passed through https://github.com/vmware-tanzu/velero/blob/master/pkg/label/label.go#L35, because the values are not completely free-form. The main limitation is length.

However, so long as that function is consistently used for writing and querying, then it should get the same backup every time.

Reply:
thanks for info. Will update PR according to this.

Reply:
updated